### DISCUSSION TITLE
[ESP32] samples blinky / gpio / custom board

### DISCUSSION THREAD

**Discussion by cederom** [zephyrproject-rtos/zephyr#36806] [Q&A] [answered]
Hello world :-)

I have a board with FPGA and additional ESP32 chip that I would like to use as FPGA controller with USB VCP terminal control. There are several problems:
* I cannot run blinky as there is no `led0` defined and no example how to define it. It would be nice to have such example. I also need to control GPIO to burn bistream into the FPGA.
* I would like to add new board to Zephyr that would allow straightforward build just using `-b my_custom_board` switch. The problem here is that `IDF_TARGET` is set to `my_custom_board` while toolchain only recognises `esp` as valid target.

Any hints welcome :-)

**Comment by b0661** [comment, marked as answer]
>  would like to add new board to Zephyr that would allow straightforward build just using -b my_custom_board switch. The problem here is that IDF_TARGET is set to my_custom_board while toolchain only recognises esp as valid target.

I had the same problem. If you copied your custom board definiton from Zephyr boards/xtensa/esp32 you may use this boards/xtensa/my_custom_board/CMakeLists.txt. It fixes IDF_TARGET and assures the Zepyhr Python version is used. I followed the [Custom Board, Devicetree and SoC definitons](https://docs.zephyrproject.org/latest/application/index.html#custom-board-devicetree-and-soc-definitions) documentation.

```
# SPDX-License-Identifier: Apache-2.0

if(CONFIG_BOOTLOADER_ESP_IDF)
  include(ExternalProject)

  ## we use hell-world project, but I think any can be used.
  set(espidf_components_dir   ${ESP_IDF_PATH}/components)
  set(espidf_prefix    ${CMAKE_BINARY_DIR}/esp-idf)
  set(espidf_build_dir ${espidf_prefix}/build)

  ExternalProject_Add(
    EspIdfBootloader
    PREFIX ${espidf_prefix}
    SOURCE_DIR ${espidf_components_dir}/bootloader/subproject
    BINARY_DIR ${espidf_build_dir}/bootloader
    CONFIGURE_COMMAND
    export PATH=$ENV{PATH}:${ESPRESSIF_TOOLCHAIN_PATH}/bin &&
    export IDF_PATH=${ESP_IDF_PATH} &&
    ${CMAKE_COMMAND} -G${CMAKE_GENERATOR}
    -S ${espidf_components_dir}/bootloader/subproject
    -B ${espidf_build_dir}/bootloader -DSDKCONFIG=${espidf_build_dir}/sdkconfig
    -DIDF_PATH=${ESP_IDF_PATH} -DIDF_TARGET=esp32
    -DPYTHON=${PYTHON_EXECUTABLE}
    -DPYTHON_DEPS_CHECKED=1
    BUILD_COMMAND
    ${CMAKE_COMMAND} --build .
    INSTALL_COMMAND ""      # This particular build system has no install command
    )

  ExternalProject_Add(
    EspPartitionTable
    SOURCE_DIR ${espidf_components_dir}/partition_table
    BINARY_DIR ${espidf_build_dir}
    CONFIGURE_COMMAND ""
    BUILD_COMMAND
    ${PYTHON_EXECUTABLE} ${ESP_IDF_PATH}/components/partition_table/gen_esp32part.py -q
    --offset 0x1000 --flash-size 4MB ${ESP_IDF_PATH}/components/partition_table/partitions_singleapp.csv ${espidf_build_dir}/partitions_singleapp.bin
    INSTALL_COMMAND ""
    )

  if(CONFIG_BUILD_OUTPUT_BIN)
    set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
      COMMAND ${PYTHON_EXECUTABLE} ${ESP_IDF_PATH}/components/esptool_py/esptool/esptool.py
      ARGS --chip esp32 elf2image --flash_mode dio --flash_freq 40m
      -o ${CMAKE_BINARY_DIR}/zephyr/${CONFIG_KERNEL_BIN_NAME}.bin
      ${CMAKE_BINARY_DIR}/zephyr/${CONFIG_KERNEL_BIN_NAME}.elf)
  endif()

  set_property(TARGET bintools PROPERTY disassembly_flag_inline_source)

  add_dependencies(app EspIdfBootloader EspPartitionTable)

  board_finalize_runner_args(esp32 "--esp-flash-bootloader=${espidf_build_dir}/bootloader/bootloader.bin")

  board_finalize_runner_args(esp32 "--esp-flash-partition_table=${espidf_build_dir}/partitions_singleapp.bin")

endif()
```

**Reply by cederom** [reply]
Allright, in fact yestreday I made it working :-) I have hardcoded `ID_TARGET=esp32` in `CMakeLists.txt` and did the `dtf` with `led0` :-) All builds and flashes fine :-) ZEPHYR ROX!! =)

**Comment by bsingh19** [comment]
Getting started with Zephyr (Setup toolchain, compilation, board bring-up, Hello World example)!!

Systems: MacBook
Board : nRF52833 DK (Optional)

https://youtu.be/4emoE-8hsqg
